### PR TITLE
fix(skore): Add tests and follow-up fixes for ConfusionMatrixDisplay

### LINKED ISSUE #2160: api(skore): follow-up fixes for #1759
### What would you like to say?

#1759 was merged and already improved by a lot the API of the `ConfusionMatrixDisplay`. However, we need still to address those little glitches:

- `report_type` is not documented in the docstring of the class while being an attribute.
- we are passing `**kwargs` in plot. However, our API is to define a variable, for instance `imshow_kwargs` which is a Python dictionary. One needs to look at the `set_style` implementation to observe that we need to have such parameter to properly set the property.
- l.156/l.157 `self.include_values = include_values` and `self.values_format = values_format` in the `plot` method. It is weird because we are not supposed to set any attribute without ending `_`. The question would be also "do we need to have those attribute stored?"
- l.285 has a local `import pandas as pd`. It should be placed on the top of the file.
- in the frame, we have a `gettattr` to get `display_labels`. It is weird because it is for sure an attribute of the class.

As another future works, I think that `_compute_data_for_display` could store a long dataframe where each row contain the number of observation, the name of the true label, the name of the predicted label (and because of #2159, certainly the estimator name and the split). It is quite aligned with the other plot. The job of `.frame()` will therefore be to reshape this long frame into a real confusion matrix. Not sure yet what would be the output of the `ComparisonReport` however.

But the first batch of comment could already be addressed.

### PR DESCRIPTION
**Pull request by GaetandeCast**
Fixes #2160, improves on #1759.

cc @glemaitre

### CODE REVIEW

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 1]
```diff
@@ -28,18 +30,38 @@ class ConfusionMatrixDisplay(DisplayMixin):
...
+        ``n_classes - 1``.
+
+    normalize : {'true', 'pred', 'all'} or None
+        Normalizes confusion matrix over the true (rows), predicted (columns)
+        conditions or all the population. If None, confusion matrix will not be
+        normalized.
+
+    report_type : {"comparison-cross-validation", "comparison-estimator", \
+            "cross-validation", "estimator"}
+        The type of report.
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: The convention that we used up to now is the same than scikit-learn. We don't document again the attributes that are just affected in `__init__`, e.g. `confusion_matrix`. Usually, we only document the attribute that are created (the one finishing with an `_`).

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 2]
```diff
@@ -96,6 +120,7 @@ def plot(
             values_format=values_format,
             cmap=cmap,
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: Let's remove the `cmap` parameter because it is actually a parameter that can be passed to `imshow_kwargs` directly. It is some old legacy in scikit-learn.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 3]
```diff
@@ -96,6 +120,7 @@ def plot(
             values_format=values_format,
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: In terms of API, I think that we should move the `value_format` into a `text_kwargs`. In short, it would be handy to benefit from the `set_style` for this part as well.

```python
display.set_style(text_kwargs=(value_format=".5g"))
display.plot()  # is going to use the format without the need to specify it again.
```

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 4]
```diff
@@ -282,9 +314,7 @@ def frame(self):
         frame : pandas.DataFrame
             The confusion matrix as a dataframe.
         """
-        import pandas as pd
-
         cm = self.confusion_matrix
-        display_labels = getattr(self, "display_labels", None)
+        display_labels = self.display_labels
 
         return pd.DataFrame(cm, index=display_labels, columns=display_labels)
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]:
```suggestion
        return pd.DataFrame(
            self.confusion_matrix,
            index=self.display_labels,
            columns=self.display_labels,
        )
```

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 2 above)
Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: The `normalize` parameter should also be added to this dict. It controls the visual that we display and does not really require any processing. Right now, we don't have a way to modify it once we instantiated the class.

Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: Oh I see that we call `sklearn_confusion_matrix` and pass the normalize. So let's never normalize the results and instead make that `plot` make the normalization. One thing we can do is to precompute in some attribute the some to be used for the different normalization.

**Review by glemaitre — COMMENTED** [review]
First thought before looking at the test

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 1 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: fixed in 27b66bc3f25dcfd8b1a6c8632c3086fda11777c4

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 2 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: done in 27b66bc3f25dcfd8b1a6c8632c3086fda11777c4 for `cmap` and in 20ddf6becba50dc6b883c7b53509b0d1054b445b for `normalize`. I also added normalize in frame.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 3 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: done in 27b66bc3f25dcfd8b1a6c8632c3086fda11777c4

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 4 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: done in c5da6364650b4494554045c768c8d6a8c1ea6d2b

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 5]
```diff
@@ -16,17 +18,16 @@ class ConfusionMatrixDisplay(DisplayMixin):
 
     Parameters
     ----------
-    confusion_matrix : ndarray of shape (n_classes, n_classes)
-        Confusion matrix.
+    confusion_matrix_data : pd.DataFrame
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: I'm not a big fan of this name tbh

Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: Yep. maybe we could just keep `confusion_matrix` directly?

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 6]
```diff
@@ -16,17 +18,16 @@ class ConfusionMatrixDisplay(DisplayMixin):
...
+        Confusion matrix data. Each row contains a confusion matrix.
 
-    display_labels : list of str, default=None
+    display_labels : list of str
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py:25]: Oh I did not realise but we don't really need this variable. `estimator.classes_` can directly be some labels with text and then we are going to reuse those strings.

Comment (glemaitre) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py:25]: Arff, you can disregard my previous comment. It is not an input from the user so fine to have them there.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 7]
```diff
@@ -139,29 +146,47 @@ def _plot_single_estimator(
...
+        cm = self.confusion_matrix_data.values.reshape(n_classes, n_classes)
+        if normalize == "true":
+            cm = cm / cm.sum(axis=1, keepdims=True)
+        elif normalize == "pred":
+            cm = cm / cm.sum(axis=0, keepdims=True)
+        elif normalize == "all":
+            cm = cm / cm.sum()
+
+        imshow_kwargs_validated = _validate_style_kwargs(
+            {"cmap": "Blues"},
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: We can put this directly in the `_default_imshow_kwargs` when we create it.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 8]
```diff
@@ -175,16 +200,28 @@ def _plot_single_estimator(
...
 
-        self.text_ = np.empty_like(cm, dtype=object)
-        if self.include_values:
-            fmt = self.values_format or (".2f" if self.normalize else "d")
+        if include_values:
+            text_kwargs_validated = _validate_style_kwargs(
+                {},
+                text_kwargs or self._default_text_kwargs or {},
+            )
+            values_format = text_kwargs_validated.pop("values_format", None)
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: We can as well add the `value_format` default in the `_default_text_kwargs` dictionary.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 9]
```diff
@@ -265,26 +297,42 @@ def _compute_data_for_display(
...
-
-        return pd.DataFrame(cm, index=display_labels, columns=display_labels)
+        n_classes = len(self.display_labels)
+        cm = self.confusion_matrix_data.values.reshape(n_classes, n_classes)
+        if normalize == "true":
+            cm = cm / cm.sum(axis=1, keepdims=True)
+        elif normalize == "pred":
+            cm = cm / cm.sum(axis=0, keepdims=True)
+        elif normalize == "all":
+            cm = cm / cm.sum()
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: Could you create a small `staticmethod` that normalize the dataframe?

Like this we can share it between the `.plot` and `.frame`

**Comment by glemaitre** [comment]
Now, I also realised that `report.metrics.confusion_matrix()` signature should be modify. We should remove `normalize` and `display_labels`. The former is only pass at `.plot` or `.frame` and the latter is not needed. Inside the function we can make that:

```
# we need to convert the classes to a list to make it hashable an used by the
# cache.
display_kwargs = {"display_labels": self._parent.estimator_.classes_.tolist()}
```

`classes_` is part of the scikit-learn API

**Comment by glemaitre** [comment]
As discussed, we can simplify a lot the plotting by using `seaborn.heatmap` instead of managing the matplotlib element ourself.

So let's refactor to have this solution. However it would imply that we need to revisit the format of the dataframe to have the following columns: "actual", "predicted", "count", "normalized_by_all", "normalized_by_predicted", "normalized_by_actual".

Then, the `.plot` and `.frame` will need to filter the right columns. For the plot when then need to make `.pivot` as shown in the seaborn documentation.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 5 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: reverted to `confusion_matrix`

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 7 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: done

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 8 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: done

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 9 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: No longer needed since we store the normalized values now, and thus only compute them once.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 10]
```diff
@@ -16,17 +17,17 @@ class ConfusionMatrixDisplay(DisplayMixin):
 
     Parameters
     ----------
-    confusion_matrix : ndarray of shape (n_classes, n_classes)
-        Confusion matrix.
+    confusion_matrix : pd.DataFrame
+        Confusion matrix data in long format with columns: "True label",
+        "Predicted label", "count", "normalized_by_true", "normalized_by_pred",
+        and "normalized_by_all". Each row represents one cell of the confusion matrix.
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py:23]: I would suggest to always use lower case and without space "true_label" and "predicted_label" when storing in the DataFrame.

At the moment of the plot, we should make the name pretty by replacing `_` by a space and capitalizing it. I don't think that we have a utility for the moment but something that I thought about in https://github.com/probabl-ai/skore/pull/1839 (still lagging behind work). For this PR, just do it locally, do not bother about the prettify function.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 11]
```diff
@@ -249,42 +184,50 @@ def _compute_data_for_display(
         cm = sklearn_confusion_matrix(
             y_true=y_true_values,
             y_pred=y_pred_values,
-            normalize=normalize,
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py:252]: small details but could you write explicitly `normalize=None` to show that we force to not have any normalization at this stage.

File: skore/tests/unit/displays/confusion_matrix/test_estimator.py [hunk 12]
```diff
@@ -77,40 +75,65 @@ def test_normalization(pyplot, forest_binary_classification_with_train_test):
...
-    custom_labels = ["Negative", "Positive"]
-    display = report.metrics.confusion_matrix(display_labels=custom_labels)
 
-    assert display.display_labels == custom_labels
-    frame = display.frame()
-    assert list(frame.index) == custom_labels
-    assert list(frame.columns) == custom_labels
+    display_true = report.metrics.confusion_matrix()
+    display_true.plot(normalize="true")
+    assert np.allclose(
```
Comment (glemaitre) [inline skore/tests/unit/displays/confusion_matrix/test_estimator.py]: Could you use `np.testing.assert_allclose` instead.

Comment (glemaitre) [reply skore/tests/unit/displays/confusion_matrix/test_estimator.py]: You can even have `np.testing.assert_allclose(arr, 1.0)` that would broadcast automatically.

File: skore/tests/unit/displays/confusion_matrix/test_estimator.py [hunk 13]
```diff
@@ -77,40 +75,65 @@ def test_normalization(pyplot, forest_binary_classification_with_train_test):
...
+        display_pred.confusion_matrix.pivot(
+            index="True label", columns="Predicted label", values="normalized_by_pred"
+        ).sum(axis=0),
+        np.ones(2),
+    )
+    assert np.allclose(display_pred.frame(normalize="pred").sum(axis=0), np.ones(2))
+
+    display_all = report.metrics.confusion_matrix()
+    display_all.plot(normalize="all")
+    assert np.isclose(
```
Comment (glemaitre) [inline skore/tests/unit/displays/confusion_matrix/test_estimator.py]: If it is a single value, I prefer:

```python
assert whatever_func() == pytest.approx(1.0)
```

It shows that we have a scalar.

**Review by glemaitre — COMMENTED** [review]
It really looks good. I only have tiny changes to make it good to be merged.
